File: ARVAR/iSNVs/recordPredictions.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# Define the file path
DF = "snvs_comb_res/ampseq_metaseq_overlap_comb_derep_decont_covFilt_97.csv"

# Set the random seed
curSeed = 42

# Read the CSV file into a DataFrame
df = pd.read_csv(DF)

# Filter out rows with NA values in Var_Al_RelPos
dfFilt = df.dropna(subset=['Var_Al_RelPos'])
dfFilt['Var_Al_RelPos'] = pd.to_numeric(dfFilt['Var_Al_RelPos'], errors='coerce')
dfFilt['Ref_Al_RelPos'] = pd.to_numeric(dfFilt['Ref_Al_RelPos'], errors='coerce')

# Set the random seed
np.random.seed(curSeed)

# Split the data into training and test sets
train_data, test_data = train_test_split(dfFilt, test_size=0.3, random_state=curSeed)

# Fit the logistic regression model
features = ['ALLELE.FREQUENCY', 'DEPTH', 'Var_Al_RelPos', 'Mean_depth']
target = 'ConsTest'

X_train = train_data[features]
y_train = train_data[target]

X_test = test_data[features]

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Predict probabilities on the test set
probs = model.predict_proba(X_test)[:, 1]

# Define the threshold
threshold = 0.5

# Create a new column for predicted values based on the threshold
test_data['PredictedConsTest'] = np.where(probs >= threshold, 1, 0)

# Calculate the number of correct predictions
equal_count = sum(test_data['ConsTest'] == test_data['PredictedConsTest'])

# Calculate the number of incorrect predictions
not_equal_count = sum(test_data['ConsTest'] != test_data['PredictedConsTest'])

# Calculate the accuracy
sumCorrect = equal_count / (equal_count + not_equal_count) * 100
print("Accuracy:", sumCorrect)

# Calculate the AUC value
roc_auc = roc_auc_score(test_data['ConsTest'], probs)
print("AUC:", roc_auc)

File: ARVAR/iSNVs/recordPredictions.R
# for this code instead of using treshold of 0.5 need to find the best treshold based on F1 score


DF = "snvs_comb_res/metaseq_ampseq_overlap_comb_derep_decont_covFilt_97.csv"

curSeed = 42

df = read.csv(DF)
dfFilt = df[!is.na(df$Var_Al_RelPos),]
dfFilt$Var_Al_RelPos = as.numeric(as.character(dfFilt$Var_Al_RelPos))
dfFilt$Ref_Al_RelPos = as.numeric(as.character(dfFilt$Ref_Al_RelPos))

set.seed(curSeed)
train_idx <- createDataPartition(dfFilt$ConsTest, p = 0.7, list = FALSE)
train_data <- dfFilt[train_idx, ]
test_data <- dfFilt[-train_idx, ]

aucModel = glm(ConsTest ~ ALLELE.FREQUENCY + STRAND.BIAS + I(DEPTH^2) + QUAL + Var_Al_RelPos + I(Var_Al_RelPos^2) + Ref_Al_RelPos + I(Ref_Al_RelPos^2), data = train_data , family = "binomial")

probs <- predict(aucModel, newdata = test_data, type = "response")

threshold <- 0.5
test_data$PredictedConsTest <- ifelse(probs >= threshold, 1, 0)

equal_count <- sum(test_data$ConsTest == test_data$PredictedConsTest)

# Count the number of times values in Column1 are NOT equal to Column2
not_equal_count <- sum(test_data$ConsTest != test_data$PredictedConsTest)

sumCorrect = equal_count / (equal_count + not_equal_count) * 100

roc_obj <- roc(test_data$ConsTest ~ probs, plot = TRUE, print.auc = TRUE)

## ampseq
DF = "snvs_comb_res/ampseq_metaseq_overlap_comb_derep_decont_covFilt_97.csv"

curSeed = 42

df = read.csv(DF)
dfFilt = df[!is.na(df$Var_Al_RelPos),]
dfFilt$Var_Al_RelPos = as.numeric(as.character(dfFilt$Var_Al_RelPos))
dfFilt$Ref_Al_RelPos = as.numeric(as.character(dfFilt$Ref_Al_RelPos))

set.seed(curSeed)
train_idx <- createDataPartition(dfFilt$ConsTest, p = 0.7, list = FALSE)
train_data <- dfFilt[train_idx, ]
test_data <- dfFilt[-train_idx, ]

aucModel = glm(ConsTest ~ ALLELE.FREQUENCY + DEPTH + Var_Al_RelPos + I(Var_Al_RelPos^2) + Mean_depth + I(Mean_depth^2), data = train_data , family = "binomial")


probs <- predict(aucModel, newdata = test_data, type = "response")

threshold <- 0.5
test_data$PredictedConsTest <- ifelse(probs >= threshold, 1, 0)

equal_count <- sum(test_data$ConsTest == test_data$PredictedConsTest)

# Count the number of times values in Column1 are NOT equal to Column2
not_equal_count <- sum(test_data$ConsTest != test_data$PredictedConsTest)

sumCorrect = equal_count / (equal_count + not_equal_count) * 100
print(sumCorrect)

roc_obj <- roc(test_data$ConsTest ~ probs, plot = TRUE, print.auc = TRUE)